File: samples-and-tests/yabe/app/models/Post.java
package models;
 
import java.util.*;
import javax.persistence.*;
 
import play.db.jpa.*;
 
@Entity
public class Post extends Model {
 
    public String title;
    public Date postedAt;
    
    @Lob
    public String content;
    
    @ManyToOne
    public User author;
    
    @OneToMany(mappedBy="post", cascade=CascadeType.ALL)
    public List<Comment> comments = new ArrayList();;
    
    public Post(User author, String title, String content) {
        this.author = author;
        this.title = title;
        this.content = content;
        this.postedAt = new Date();
    }
    
    public Post addComment(String author, String content) {
        Comment newComment = new Comment(this, author, content).save();
        this.comments.add(newComment);
        return this;
    }
    
    public Post previous() {
        return Post.find("postedAt < ? order by postedAt desc", postedAt).first();
    }

    public Post next() {
        return Post.find("postedAt > ? order by postedAt asc", postedAt).first();
    }
 
}

File: samples-and-tests/yabe-with-scala/app/models/Post.scala
package models
 
import java.util._
import javax.persistence._
 
import play.db.jpa._
import play.data.validation._
 
@Entity
class Post(

    @Required
    @ManyToOne
    var author: User,
    
    @Required
    var title: String,
    
    @Lob
    @Required
    @MaxSize(10000)
    var content: String

) extends Model[Post] {
    
    @Required
    var postedAt = new Date()  
    
    @OneToMany(mappedBy="post", cascade=Array(CascadeType.ALL))
    var comments: List[Comment] = new ArrayList[Comment]
    
    @ManyToMany(cascade=Array(CascadeType.PERSIST))
    var tags: Set[Tag] = new TreeSet[Tag]
    
    def addComment(author: String, content: String) = {
        val newComment = new Comment(this, author, content)
        newComment.save  
        comments.add(newComment)
        this
    }
    
    def previous = {
        Post.find("postedAt < ? order by postedAt desc", postedAt).first
    }

    def next = {
        Post.find("postedAt > ? order by postedAt asc", postedAt).first
    }
    
    def tagItWith(name: String) = {
        tags add Tag.findOrCreateByName(name)
        this
    }
    
    override def toString() = {
        title
    }
 
}

object Post extends Model[Post]{
    
    def findTaggedWith(tag: String) = {
        Post.find("select distinct p from Post p join p.tags as t where t.name = ?", tag).fetch
    }
    
    def findTaggedWith(tags: String*) = {
        Nil
    }
    
}
